### DISCUSSION TITLE
Is KafkaTopicOperator enabled by default?

### DISCUSSION THREAD

**Discussion by jslusher** [strimzi/strimzi-kafka-operator#10707] [Q&A] [answered]
I have a strimzi Kafka cluster and I can see several KafkaTopics in the output of `kubectl get kafkatopic`, but I don't have the topic operator enabled as far as I know. I don't see a topic-operator pod either. Is it enabled whether I want it or not?

**Comment by scholzj** [comment, marked as answer]
The Topic Operator is enabled by having the `.spec.entityOperator.topicOperator` section in the Kafka CR. Having it there will have the Topic Operator start (the pod will be actually named `...-entity-operator` and the Topic Operator is just one of its containers).

But the KafkaTopic CRD is installed independently on it with the Strimzi Cluster Operator. So even if you haven't enabled the Topic Operator, you might create the `KafkaTopic` resources. But they will be ignored as there will be nobody to _pick them up_.

**Reply by jslusher** [reply]
I see now that I have `.spec.entityOperator.topicOperator` configured with logging and resources in my `Kafka` resource. I missed when I looked before because I was looking for `topic-operator`.  I might have also been looking for it in my operator config and not on the Kafka resource itself. Anyway, found it, thanks!

**Comment by jslusher** [comment]
@scholzj I have a followup question about this. My strimzi operator started around version 0.23.0 or so, which means originally we were using the BTO. We're now using version 0.42.0. I see this in [your UTO blog post](https://strimzi.io/blog/2023/11/02/unidirectional-topic-operator/), in the section on upgrading to UTO from BTO:
> The plan is to move it to beta in Strimzi 0.39.0, where it will be enabled by default.

Does that mean that we're already using UTO or do I still need to set `STRIMZI_FEATURE_GATES="+UnidirectionalTopicOperator"` to make the switch?

I'm also wondering if I should just remove the topic-operator altogether if we don't plan on using the `KafkaTopic` resource. Currently we do not have any `KafkaTopic` resources that weren't created automatically by the operator.

**Reply by scholzj** [reply]
Yes, if you use Strimzi 0.42.0, you are already using the Unidirectional Topic Operator by default (in fact, the Bidirectional Topic Operator is already removed before 0.42.0).

If you do not plan to use the KafkaTopic resources, disabling it might be the best option. It should not cause any harm in general ... but you would save some compute resources and so on.

**Reply by jslusher** [reply]
What happens to the existing KafkaTopics if the toipc-operator is disabled?

**Reply by scholzj** [reply]
Once you disable it (and once the Pod / Container is deleted!), they would be ignored. But you should probably consider deleting them I guess as they would be no longer needed.

**Reply by jslusher** [reply]
I removed my topic-operator from my staging environment and I verified the container is no longer present in the entity-operator pod. I tried deleting one of the KafkaTopics and it seems to be hanging. I don't see anything in the events or in the operator logs. I'm not siure what's going on there.

```
kubectl delete kt strimzi.cruisecontrol.metrics --context staging -n strimzi-kafka-cluster
kafkatopic.kafka.strimzi.io "strimzi.cruisecontrol.metrics" deleted
(hanging)
```

**Reply by scholzj** [reply]
Hmm, you are right ... the topics still have the finalizers. I guess you can try a command like this to remove them first:

```
kubectl get kt -o=json | jq '.items[].metadata.finalizers = null' | kubectl apply -f -
```

Other options might be to do it manually topic by topic with `kubect edit`. Or you can reenable the Topic Operator and:
* Either unmanage the topics and delete them: https://strimzi.io/docs/operators/latest/full/deploying.html#proc-converting-managed-topics-str
* or disable the finalizers: https://strimzi.io/docs/operators/latest/full/deploying.html#deploying-the-topic-operator-standalone-str

**Reply by jslusher** [reply]
The second option, which is what I was leaning toward, isn't going to work because my topicOperator isn't standalone. I don't see an option to set env in `kafka.spec.entityOperator.topicOperator`. Am I correct about that? If so I'll go with the first option I think. I've heard that editing finalizers directly out of objects can cause issues.

**Reply by scholzj** [reply]
You can use the https://strimzi.io/docs/operators/latest/full/configuring.html#type-ContainerTemplate-reference to specify the ENV var for the TO container. It will be under `.spec.entityOperator.template.topicOperatorContainer` or something like that.
